File: sdk/python/src/a3s_box/code_interpreter.py
"""Small local Code Interpreter facade built on the native Sandbox API."""

from __future__ import annotations

from collections.abc import Mapping

from .models import CommandResult
from .runtime import AsyncLocalRuntime, LocalRuntime
from .sandbox import AsyncSandbox as BaseAsyncSandbox
from .sandbox import Sandbox as BaseSandbox

DEFAULT_CODE_INTERPRETER_IMAGE = "python:3.12-alpine"


class Sandbox(BaseSandbox):
    @classmethod
    def create(
        cls,
        template: str | None = None,
        *,
        timeout: int = 3600,
        envs: Mapping[str, str] | None = None,
        metadata: Mapping[str, str] | None = None,
        runtime: LocalRuntime | None = None,
    ) -> Sandbox:
        return super().create(
            template or DEFAULT_CODE_INTERPRETER_IMAGE,
            timeout=timeout,
            envs=envs,
            metadata=metadata,
            runtime=runtime,
        )

    def run_code(
        self,
        code: str,
        *,
        language: str = "python",
        timeout: float | None = None,
    ) -> CommandResult:
        if language != "python":
            raise ValueError("the local Code Interpreter currently supports Python only")
        return self.commands.run(["python", "-c", code], timeout=timeout)


class AsyncSandbox(BaseAsyncSandbox):
    @classmethod
    async def create(
        cls,
        template: str | None = None,
        *,
        timeout: int = 3600,
        envs: Mapping[str, str] | None = None,
        metadata: Mapping[str, str] | None = None,
        runtime: AsyncLocalRuntime | None = None,
    ) -> AsyncSandbox:
        return await super().create(
            template or DEFAULT_CODE_INTERPRETER_IMAGE,
            timeout=timeout,
            envs=envs,
            metadata=metadata,
            runtime=runtime,
        )

    async def run_code(
        self,
        code: str,
        *,
        language: str = "python",
        timeout: float | None = None,
    ) -> CommandResult:
        if language != "python":
            raise ValueError("the local Code Interpreter currently supports Python only")
        return await self.commands.run(["python", "-c", code], timeout=timeout)


__all__ = [
    "AsyncSandbox",
    "DEFAULT_CODE_INTERPRETER_IMAGE",
    "Sandbox",
]

File: sdk/typescript/src/code-interpreter.ts
import {
  Sandbox as BaseSandbox,
  type CommandResult,
  type SandboxCreateOptions,
} from './sandbox.js'

export const DEFAULT_CODE_INTERPRETER_IMAGE = 'python:3.12-alpine'

/** Minimal local Python interpreter facade over the native Sandbox API. */
export class Sandbox {
  private constructor(private readonly sandbox: BaseSandbox) {}

  static async create(
    template = DEFAULT_CODE_INTERPRETER_IMAGE,
    options: SandboxCreateOptions = {}
  ): Promise<Sandbox> {
    return new Sandbox(await BaseSandbox.create(template, options))
  }

  get sandboxId(): string {
    return this.sandbox.sandboxId
  }

  get id(): string {
    return this.sandbox.id
  }

  get commands(): BaseSandbox['commands'] {
    return this.sandbox.commands
  }

  get files(): BaseSandbox['files'] {
    return this.sandbox.files
  }

  async runCode(
    code: string,
    options: { language?: 'python'; timeoutMs?: number } = {}
  ): Promise<CommandResult> {
    if (options.language !== undefined && options.language !== 'python') {
      throw new Error('the local Code Interpreter currently supports Python only')
    }
    return this.commands.run(['python', '-c', code], {
      timeoutMs: options.timeoutMs,
    })
  }

  kill(): Promise<void> {
    return this.sandbox.kill()
  }

  pause(options: { keepMemory?: boolean } = {}): Promise<void> {
    return this.sandbox.pause(options)
  }

  resume(): Promise<void> {
    return this.sandbox.resume()
  }

  isRunning(): Promise<boolean> {
    return this.sandbox.isRunning()
  }
}

export default Sandbox
